File: src/co/nums/intellij/aem/htl/completion/provider/inserthandlers/HtlExprOptionBracketsInsertHandler.java
package co.nums.intellij.aem.htl.completion.provider.inserthandlers;

import com.intellij.codeInsight.completion.InsertHandler;
import com.intellij.codeInsight.completion.InsertionContext;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.ScrollType;
import com.intellij.util.text.CharArrayUtil;

public class HtlExprOptionBracketsInsertHandler implements InsertHandler<LookupElement> {

	@Override
	public void handleInsert(InsertionContext context, LookupElement item) {
		Editor editor = context.getEditor();
		Document document = editor.getDocument();
		int caretOffset = editor.getCaretModel().getOffset();
		if (!hasBrackets(document, caretOffset)) {
			insertBrackets(context, document, caretOffset);
			moveCaretIntoBrackets(editor, caretOffset);
		}
	}

	private static boolean hasBrackets(Document document, int caretOffset) {
		CharSequence chars = document.getCharsSequence();
		return CharArrayUtil.regionMatches(chars, caretOffset, "=[");
	}

	private static void insertBrackets(InsertionContext context, Document document, int caretOffset) {
		document.insertString(caretOffset, "=[]");
		if (context.getCompletionChar() == '=') {
			context.setAddCompletionChar(false); // IDEA-19449
		}
	}

	private static void moveCaretIntoBrackets(Editor editor, int caretOffset) {
		editor.getCaretModel().moveToOffset(caretOffset + 2);
		editor.getScrollingModel().scrollToCaret(ScrollType.RELATIVE);
		editor.getSelectionModel().removeSelection();
	}

}

File: src/co/nums/intellij/aem/htl/completion/provider/inserthandlers/HtlExprOptionBracketsInsertHandler.kt
package co.nums.intellij.aem.htl.completion.provider.inserthandlers

import co.nums.intellij.aem.extensions.hasText
import co.nums.intellij.aem.extensions.moveCaret
import com.intellij.codeInsight.completion.InsertHandler
import com.intellij.codeInsight.completion.InsertionContext
import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.openapi.editor.Document

object HtlExprOptionBracketsInsertHandler : InsertHandler<LookupElement> {

    private const val INTO_BRACKETS_OFFSET = 2

    override fun handleInsert(context: InsertionContext, item: LookupElement) {
        val editor = context.editor
        val document = editor.document
        val offset = editor.caretModel.offset
        if (!hasBrackets(document, offset)) {
            insertBrackets(context, document, offset)
            editor.moveCaret(INTO_BRACKETS_OFFSET)
        }
    }

    private fun hasBrackets(document: Document, offset: Int) = document.hasText(offset, "=[")

    private fun insertBrackets(context: InsertionContext, document: Document, offset: Int) {
        document.insertString(offset, "=[]")
        if (context.completionChar == '=') {
            context.setAddCompletionChar(false) // IDEA-19449
        }
    }

}
